core/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -74,4 +112,86 @@ public Task unregister(Task task) {
     public Map<Long, Task> getTasks() {
         return Collections.unmodifiableMap(new HashMap<>(tasks));
     }
+
+    /**
+     * Returns the number of currently banned tasks.
+     *
+     * Will be used in task manager stats and for debugging.
+     */
+    public int banCount() {
+        return banedParents.size();
+    }
+
+    /**
+     * Bans all tasks with the specified parent task from execution, cancels all tasks that are currently executing.
+     *
+     * This method is called when a parent task that has children is cancelled.
+     */
+    public void setBan(String parentNode, long parentId, String reason) {
+        // Set the ban first, so the newly created tasks cannot be registered
+        Tuple<String, Long> ban = new Tuple<>(parentNode, parentId);
+        synchronized (banedParents) {
+            if (lastDiscoveryNodes == null || lastDiscoveryNodes.nodeExists(parentNode)) {
+                // Only set the ban if the node is the part of the cluster
+                banedParents.put(ban, reason);
+            }
+        }
+
+        // Now go through already running tasks and cancel them
+        for (Map.Entry<Long, Task> taskEntry : tasks.entrySet()) {
+            Task task = taskEntry.getValue();
+            if (parentId == task.getParentId() && parentNode.equals(task.getParentNode())) {
+                assert task instanceof CancellableTask;
```

Comment:
I think that assert is pretty unnecessary since you get a much better exception if you let it run into the classcast below? I think you should make the cast outside of the catch block instead and let if it fail hard all the time?